### PR TITLE
Expand tilde in resource file path

### PR DESCRIPTION
**Pull request by mkienow-r7**
If the resource file path starts with a "~" it will be expanded to the process owner's home directory (the environment variable HOME must be set correctly). Calling the `resource` command with a path that started with a tilde would previously result in an error that the "$FILENAME is not a valid resource file".

## Verification

- [x] Create a simple resource file: `echo "workspace -v" > ~/Downloads/res_test.rc`
- [x] Start `msfconsole`
- [x] Run the command `resource ~/Downloads/res_test.rc`
- [x] **Verify** that the resource script commands are run and that workspaces are listed verbosely
- [x] Run the `resource` command again using the absolute path to the resource file
- [x] **Verify** that the resource script commands are run and that workspaces are listed verbosely

### CODE REVIEW

File: lib/msf/ui/console/command_dispatcher/resource.rb [hunk 1]
```diff
@@ -51,11 +51,12 @@ def cmd_resource(*args)
...
             args.each do |res|
+              res_expand = ::File.expand_path(res)
               good_res = nil
               if res == '-'
                 good_res = res
-              elsif ::File.exist?(res)
-                good_res = res
+              elsif ::File.exist?(res_expand)
+                good_res = res_expand
               elsif
```
Comment (wvu-r7) [inline lib/msf/ui/console/command_dispatcher/resource.rb]: This condition-less `elsif` is bugging me, even though it's not your code.

Comment (wvu-r7) [inline lib/msf/ui/console/command_dispatcher/resource.rb]: This condition-less `elsif` is bugging me, even though it's not your code.

**Comment by wvu-r7** [comment]
Please update `resource` tab completion to handle tilde expansion.

A couple other things (they were already broken):

```
msf5 > resource ""
[-] Error while running command resource: Is a directory @ io_fread - /home/wvu/metasploit-framework/scripts/resource/

Call stack:
/home/wvu/metasploit-framework/lib/msf/ui/console/driver.rb:269:in `read'
/home/wvu/metasploit-framework/lib/msf/ui/console/driver.rb:269:in `load_resource'
/home/wvu/metasploit-framework/lib/msf/ui/console/command_dispatcher/resource.rb:73:in `block in cmd_resource'
/home/wvu/metasploit-framework/lib/msf/ui/console/command_dispatcher/resource.rb:53:in `each'
/home/wvu/metasploit-framework/lib/msf/ui/console/command_dispatcher/resource.rb:53:in `cmd_resource'
/home/wvu/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:546:in `run_command'
/home/wvu/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:508:in `block in run_single'
/home/wvu/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:502:in `each'
/home/wvu/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:502:in `run_single'
/home/wvu/metasploit-framework/lib/rex/ui/text/shell.rb:208:in `run'
/home/wvu/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
/home/wvu/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'
./msfconsole:49:in `<main>'
msf5 > resource .
[-] Error while running command resource: Is a directory @ io_fread - .

Call stack:
/home/wvu/metasploit-framework/lib/msf/ui/console/driver.rb:269:in `read'
/home/wvu/metasploit-framework/lib/msf/ui/console/driver.rb:269:in `load_resource'
/home/wvu/metasploit-framework/lib/msf/ui/console/command_dispatcher/resource.rb:73:in `block in cmd_resource'
/home/wvu/metasploit-framework/lib/msf/ui/console/command_dispatcher/resource.rb:53:in `each'
/home/wvu/metasploit-framework/lib/msf/ui/console/command_dispatcher/resource.rb:53:in `cmd_resource'
/home/wvu/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:546:in `run_command'
/home/wvu/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:508:in `block in run_single'
/home/wvu/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:502:in `each'
/home/wvu/metasploit-framework/lib/rex/ui/text/dispatcher_shell.rb:502:in `run_single'
/home/wvu/metasploit-framework/lib/rex/ui/text/shell.rb:208:in `run'
/home/wvu/metasploit-framework/lib/metasploit/framework/command/console.rb:48:in `start'
/home/wvu/metasploit-framework/lib/metasploit/framework/command/base.rb:82:in `start'
./msfconsole:49:in `<main>'
msf5 >
```

Empties and dirs drop an ugly call stack.

File: lib/msf/ui/console/command_dispatcher/resource.rb (hunk 1 above)
Comment (mkienow-r7) [reply lib/msf/ui/console/command_dispatcher/resource.rb]: I'll change that, one minute...

**Comment by mkienow-r7** [comment]
@wvu-r7 The `resource` tab completion is enhanced to support tilde expansion, and I made the file checks more robust to fix the preexisting issues you pointed out.

```
msf5 > resource .
[-] . is not a valid resource file
msf5 > resource ""
[-]  is not a valid resource file
msf5 > resource ~
[-] ~ is not a valid resource file
msf5 > resource ~/Downloads
[-] ~/Downloads is not a valid resource file
```

**Comment by wvu-r7** [comment]
# Release Notes

This adds tilde expansion to the `resource` command and its tab completion.

**Pull request merged**